File: 0056-merge-intervals/0056-merge-intervals.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        vector<vector<int>> answer; 
        sort(intervals.begin(), intervals.end()); 
        int prev = intervals[0][1]; 
        answer.push_back(intervals[0]);
        for(int i = 1; i < intervals.size(); i++){
            int next = intervals[i][0]; 
            if(prev >= next){
                prev = max(prev,intervals[i][1]); 
                answer[answer.size()-1][1] = max(answer[answer.size()-1][1], intervals[i][1]); 
            } else{
                prev = intervals[i][1]; 
                answer.push_back(intervals[i]); 
            }
        }
        return answer; 
    }
};

File: 0056-merge-intervals/0056-merge-intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, (a,b) -> a[0] - b[0]); 
        List<int[]> answer = new ArrayList<>(); 
        int end = 0, n = intervals.length; 
        while(end < n){
            int[] curr = intervals[end]; 
            int nextMax = curr[1]; 
            while(end + 1 < n && intervals[end+1][0] <= nextMax){
                nextMax = Math.max(nextMax, intervals[end+1][1]); 
                end++; 
            }

            answer.add(new int[]{curr[0], nextMax}); 
            end++; 
        }
        
        return answer.toArray(new int[answer.size()][]);
    }
}